src/libPMacc/include/eventSystem/events/kernelEvents.hpp

```diff
@@ -41,43 +46,243 @@
+            m_blockExtent( blockExtent ),
+            m_sharedMemByte( sharedMemByte )
+        {
+
+        }
+
+        /** execute the kernel functor
+         * 
+         * @tparam T_Args types of the arguments
+         * @param args arguments for the kernel functor
+         * 
+         * @{
+         */
+        template<
+            typename ... T_Args
+        >
+        HINLINE
+        void
+        operator()(
+            T_Args const & ... args
+        ) const
+        {
+
+            std::string const kernelName = typeid( m_kernel.m_kernelFunctor ).name();
+            std::string const kernelInfo = kernelName +
+                std::string( " [" ) + m_kernel.m_file + std::string( ":" ) + 
+                std::to_string( m_kernel.m_line ) + std::string( " ]" ); 
+
+            CUDA_CHECK_KERNEL_MSG(
+                cudaDeviceSynchronize( ),
+                std::string( "Crash before kernel call " ) + kernelInfo
+            );
+
+            PMacc::TaskKernel* taskKernel = PMacc::Environment<>::get().Factory().createTaskKernel(
+                typeid( kernelName ).name()
+            );
+
+            DataSpace<
+                traits::GetNComponents<
+                    T_VectorGrid
+                >::value
+            > gridExtent( m_gridExtent );
+            
+            DataSpace<
+                traits::GetNComponents<
+                    T_VectorBlock
+                >::value
+            > blockExtent( m_blockExtent );
+            
+            nvidia::gpuEntryFunction<<<
+                gridExtent,
+                blockExtent,
+                m_sharedMemByte,
+                taskKernel->getCudaStream()
+            >>>( 
+                m_kernel.m_kernelFunctor, 
+                args ... 
+            );
+            CUDA_CHECK_KERNEL_MSG(
+                cudaGetLastError( ),
+                std::string( "Last error after kernel launch " ) + kernelInfo
+            );
+            CUDA_CHECK_KERNEL_MSG(
+                cudaDeviceSynchronize( ),
+                std::string( "Crash after kernel launch " ) + kernelInfo
+            );
+            taskKernel->activateChecks( );
+            CUDA_CHECK_KERNEL_MSG(
+                cudaDeviceSynchronize( ),
+                std::string(  "Crash after kernel activation" ) + kernelInfo
+            );
+        }
+        
+        template<
+            typename ... T_Args
+        >
+        HINLINE
+        void
+        operator()(
+            T_Args const &... args
+        ) 
+        {
+            return static_cast< const KernelStarter >(*this)( args ... );
+        }
+        
+        /** @} */
+
+    };
+
+
+    /** creates a kernel object
+     * 
+     * @tparam T_KernelFunctor type of the kernel functor
+     * @param kernelFunctor instance of the functor
+     * @param file file name (for debug)
+     * @param line line number in the file (for debug)
+     */
+    template< typename T_KernelFunctor >
+    auto kernel(
```

Comment:
Mybe we should move this function and all struct above to a unique namespace inside `PMacc` else we can't use the name `kernel` as namespace inside PMacc.